File: android/src/main/java/com/swipepredictor/SwipePredictorModule.kt
package com.swipepredictor

import android.os.Handler
import android.os.HandlerThread
import android.view.Choreographer
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.roundToLong

class SwipePredictorModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {
    
    companion object {
        init {
            System.loadLibrary("swipe_predictor")
        }
        
        // Native method declarations
        @JvmStatic
        external fun nativeInitManager(
            decelerationRate: Double,
            minVelocityThreshold: Double,
            minGestureTimeMs: Double,
            velocitySmoothingFactor: Double
        )
        
        @JvmStatic
        external fun nativeInitPredictor(): Int
        
        @JvmStatic
        external fun nativeAddTouchPoint(predictorId: Int, x: Double, y: Double, timestamp: Double)
        
        @JvmStatic
        external fun nativeGetPrediction(predictorId: Int): Prediction?
        
        @JvmStatic
        external fun nativeResetPredictor(predictorId: Int)
        
        @JvmStatic
        external fun nativeDetectCancellation(predictorId: Int): Int
        
        @JvmStatic
        external fun nativeRemovePredictor(predictorId: Int)
    }
    
    private val handlerThread = HandlerThread("SwipePredictorThread").apply { start() }
    private val backgroundHandler = Handler(handlerThread.looper)
    private val activePredictors = ConcurrentHashMap<Int, PredictorInfo>()
    private var choreographer: Choreographer? = null
    private var isFrameCallbackScheduled = false
    
    data class PredictorInfo(
        val predictorId: Int,
        var lastUpdate: Long = 0,
        val updateInterval: Long,
        val confidenceThreshold: Double
    )
    
    init {
        // Initialize the Rust predictor manager with default physics
        nativeInitManager(1500.0, 50.0, 30.0, 0.7)
    }
    
    override fun getName() = "SwipePredictorModule"
    
    @ReactMethod
    fun createPredictor(options: ReadableMap, promise: Promise) {
        backgroundHandler.post {
            try {
                val predictorId = nativeInitPredictor()
                
                if (predictorId >= 0) {
                    val updateInterval = (options.getDouble("updateInterval") * 1000).roundToLong()
                        .coerceAtLeast(8) // Min 8ms (120fps)
                    val confidenceThreshold = options.getDouble("confidenceThreshold")
                    
                    val info = PredictorInfo(
                        predictorId = predictorId,
                        updateInterval = updateInterval,
                        confidenceThreshold = confidenceThreshold
                    )
                    
                    activePredictors[predictorId] = info
                    
                    reactApplicationContext.runOnUiQueueThread {
                        startChoreographerIfNeeded()
                    }
                    
                    promise.resolve(predictorId)
                } else {
                    promise.reject("INIT_FAILED", "Failed to initialize predictor")
                }
            } catch (e: Exception) {
                promise.reject("INIT_ERROR", "Error initializing predictor", e)
            }
        }
    }
    
    @ReactMethod
    fun addTouchPoint(predictorId: Int, x: Double, y: Double, timestamp: Double) {
        backgroundHandler.post {
            nativeAddTouchPoint(predictorId, x, y, timestamp)
        }
    }
    
    @ReactMethod
    fun getPrediction(predictorId: Int, promise: Promise) {
        backgroundHandler.post {
            try {
                val prediction = nativeGetPrediction(predictorId)
                val info = activePredictors[predictorId]
                
                if (prediction != null && info != null && prediction.confidence >= info.confidenceThreshold) {
                    val result = Arguments.createMap().apply {
                        putDouble("x", prediction.x)
                        putDouble("y", prediction.y)
                        putDouble("confidence", prediction.confidence)
                    }
                    promise.resolve(result)
                } else {
                    promise.resolve(null)
                }
            } catch (e: Exception) {
                promise.reject("PREDICTION_ERROR", "Error getting prediction", e)
            }
        }
    }
    
    @ReactMethod
    fun resetPredictor(predictorId: Int) {
        backgroundHandler.post {
            nativeResetPredictor(predictorId)
        }
    }
    
    @ReactMethod
    fun detectCancellation(predictorId: Int, promise: Promise) {
        backgroundHandler.post {
            try {
                val isCancelled = nativeDetectCancellation(predictorId) == 1
                promise.resolve(isCancelled)
            } catch (e: Exception) {
                promise.reject("CANCELLATION_ERROR", "Error detecting cancellation", e)
            }
        }
    }
    
    @ReactMethod
    fun removePredictor(predictorId: Int) {
        backgroundHandler.post {
            nativeRemovePredictor(predictorId)
            activePredictors.remove(predictorId)
            
            if (activePredictors.isEmpty()) {
                reactApplicationContext.runOnUiQueueThread {
                    stopChoreographer()
                }
            }
        }
    }
    
    private fun startChoreographerIfNeeded() {
        if (choreographer == null) {
            choreographer = Choreographer.getInstance()
        }
        
        if (!isFrameCallbackScheduled) {
            isFrameCallbackScheduled = true
            choreographer?.postFrameCallback(frameCallback)
        }
    }
    
    private fun stopChoreographer() {
        isFrameCallbackScheduled = false
        choreographer?.removeFrameCallback(frameCallback)
    }
    
    private val frameCallback = object : Choreographer.FrameCallback {
        override fun doFrame(frameTimeNanos: Long) {
            if (!isFrameCallbackScheduled) return
            
            val currentTime = System.currentTimeMillis()
            
            activePredictors.forEach { (predictorId, info) ->
                if (currentTime - info.lastUpdate >= info.updateInterval) {
                    checkAndSendPrediction(predictorId)
                    info.lastUpdate = currentTime
                }
            }
            
            // Schedule next frame
            choreographer?.postFrameCallback(this)
        }
    }
    
    private fun checkAndSendPrediction(predictorId: Int) {
        backgroundHandler.post {
            try {
                val prediction = nativeGetPrediction(predictorId)
                val info = activePredictors[predictorId]
                
                if (prediction != null && info != null && prediction.confidence >= info.confidenceThreshold) {
                    val params = Arguments.createMap().apply {
                        putInt("predictorId", predictorId)
                        putDouble("x", prediction.x)
                        putDouble("y", prediction.y)
                        putDouble("confidence", prediction.confidence)
                    }
                    
                    sendEvent("onPrediction", params)
                }
                
                // Check for cancellation
                val isCancelled = nativeDetectCancellation(predictorId) == 1
                if (isCancelled) {
                    val params = Arguments.createMap().apply {
                        putInt("predictorId", predictorId)
                    }
                    sendEvent("onCancellation", params)
                }
            } catch (e: Exception) {
                // Log error but don't crash
                e.printStackTrace()
            }
        }
    }
    
    private fun sendEvent(eventName: String, params: WritableMap) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }
    
    override fun onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy()
        handlerThread.quitSafely()
        stopChoreographer()
    }
}

File: ios/SwipePredictor/SwipePredictorModule.swift
import Foundation
import React

@objc(SwipePredictorModule)
class SwipePredictorModule: RCTEventEmitter {
    private var predictorUpdateTimer: Timer?
    private var activePredictors: [Int: PredictorInfo] = [:]
    private var displayLink: CADisplayLink?
    
    struct PredictorInfo {
        let predictorId: Int32
        var lastUpdate: Date
        var updateInterval: TimeInterval
        var confidenceThreshold: Double
        var onPredictionCallback: String?
    }
    
    override init() {
        super.init()
        
        // Initialize the Rust predictor manager with default physics
        init_predictor_manager(1500.0, 50.0, 30.0, 0.7)
    }
    
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    override func supportedEvents() -> [String]! {
        return ["onPrediction", "onCancellation"]
    }
    
    @objc
    func createPredictor(_ options: [String: Any], 
                        resolver resolve: @escaping RCTPromiseResolveBlock,
                        rejecter reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.global(qos: .userInteractive).async { [weak self] in
            let predictorId = init_predictor()
            
            if predictorId >= 0 {
                let updateInterval = (options["updateInterval"] as? TimeInterval) ?? 0.016 // 60 FPS default
                let confidenceThreshold = (options["confidenceThreshold"] as? Double) ?? 0.7
                let onPredictionCallback = options["onPredictionCallback"] as? String
                
                let info = PredictorInfo(
                    predictorId: predictorId,
                    lastUpdate: Date(),
                    updateInterval: updateInterval,
                    confidenceThreshold: confidenceThreshold,
                    onPredictionCallback: onPredictionCallback
                )
                
                DispatchQueue.main.async {
                    self?.activePredictors[Int(predictorId)] = info
                    self?.startDisplayLinkIfNeeded()
                    resolve(predictorId)
                }
            } else {
                reject("INIT_FAILED", "Failed to initialize predictor", nil)
            }
        }
    }
    
    @objc
    func addTouchPoint(_ predictorId: Int,
                      x: Double,
                      y: Double,
                      timestamp: Double) {
        DispatchQueue.global(qos: .userInteractive).async {
            add_touch_point(Int32(predictorId), x, y, timestamp)
        }
    }
    
    @objc
    func getPrediction(_ predictorId: Int,
                      resolver resolve: @escaping RCTPromiseResolveBlock,
                      rejecter reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.global(qos: .userInteractive).async { [weak self] in
            var x: Double = 0.0
            var y: Double = 0.0
            var confidence: Double = 0.0
            
            let result = get_prediction(Int32(predictorId), &x, &y, &confidence)
            
            if result == 1 {
                let prediction: [String: Any] = [
                    "x": x,
                    "y": y,
                    "confidence": confidence
                ]
                
                DispatchQueue.main.async {
                    if let info = self?.activePredictors[predictorId],
                       confidence >= info.confidenceThreshold {
                        resolve(prediction)
                    } else {
                        resolve(nil)
                    }
                }
            } else {
                DispatchQueue.main.async {
                    resolve(nil)
                }
            }
        }
    }
    
    @objc
    func resetPredictor(_ predictorId: Int) {
        DispatchQueue.global(qos: .userInteractive).async {
            reset_predictor(Int32(predictorId))
        }
    }
    
    @objc
    func detectCancellation(_ predictorId: Int,
                           resolver resolve: @escaping RCTPromiseResolveBlock,
                           rejecter reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.global(qos: .userInteractive).async {
            let isCancelled = detect_cancellation(Int32(predictorId))
            resolve(isCancelled == 1)
        }
    }
    
    @objc
    func removePredictor(_ predictorId: Int) {
        DispatchQueue.global(qos: .userInteractive).async { [weak self] in
            remove_predictor(Int32(predictorId))
            
            DispatchQueue.main.async {
                self?.activePredictors.removeValue(forKey: predictorId)
                if self?.activePredictors.isEmpty == true {
                    self?.stopDisplayLink()
                }
            }
        }
    }
    
    private func startDisplayLinkIfNeeded() {
        guard displayLink == nil else { return }
        
        displayLink = CADisplayLink(target: self, selector: #selector(displayLinkUpdate))
        displayLink?.preferredFramesPerSecond = 120 // Support ProMotion displays
        displayLink?.add(to: .current, forMode: .common)
    }
    
    private func stopDisplayLink() {
        displayLink?.invalidate()
        displayLink = nil
    }
    
    @objc
    private func displayLinkUpdate() {
        let now = Date()
        
        for (predictorId, info) in activePredictors {
            if now.timeIntervalSince(info.lastUpdate) >= info.updateInterval {
                checkAndSendPrediction(predictorId: predictorId)
                activePredictors[predictorId]?.lastUpdate = now
            }
        }
    }
    
    private func checkAndSendPrediction(predictorId: Int) {
        DispatchQueue.global(qos: .userInteractive).async { [weak self] in
            var x: Double = 0.0
            var y: Double = 0.0
            var confidence: Double = 0.0
            
            let result = get_prediction(Int32(predictorId), &x, &y, &confidence)
            
            if result == 1 {
                DispatchQueue.main.async {
                    if let info = self?.activePredictors[predictorId],
                       confidence >= info.confidenceThreshold {
                        let prediction: [String: Any] = [
                            "predictorId": predictorId,
                            "x": x,
                            "y": y,
                            "confidence": confidence
                        ]
                        
                        self?.sendEvent(withName: "onPrediction", body: prediction)
                    }
                }
            }
            
            // Check for cancellation
            let isCancelled = detect_cancellation(Int32(predictorId))
            if isCancelled == 1 {
                DispatchQueue.main.async {
                    self?.sendEvent(withName: "onCancellation", body: ["predictorId": predictorId])
                }
            }
        }
    }
}

// Rust FFI function declarations
@_silgen_name("init_predictor_manager")
func init_predictor_manager(_ deceleration_rate: Double, 
                           _ min_velocity_threshold: Double,
                           _ min_gesture_time_ms: Double,
                           _ velocity_smoothing_factor: Double)

@_silgen_name("init_predictor")
func init_predictor() -> Int32

@_silgen_name("add_touch_point")
func add_touch_point(_ predictor_id: Int32, _ x: Double, _ y: Double, _ timestamp: Double)

@_silgen_name("get_prediction")
func get_prediction(_ predictor_id: Int32, 
                   _ out_x: UnsafeMutablePointer<Double>,
                   _ out_y: UnsafeMutablePointer<Double>,
                   _ out_confidence: UnsafeMutablePointer<Double>) -> Int32

@_silgen_name("reset_predictor")
func reset_predictor(_ predictor_id: Int32)

@_silgen_name("detect_cancellation")
func detect_cancellation(_ predictor_id: Int32) -> Int32

@_silgen_name("remove_predictor")
func remove_predictor(_ predictor_id: Int32)